### ISSUE TITLE
Model Fails Statics Checks

### ISSUE THREAD

**Issue by bjhowie** [JWock82/PyNite#102]
Sorry to come again with problems and not solutions, but i'm having some trouble with a proof-of-concept lateral stability model that i have been trying to analyse with PyNite. It is a model of an RC building core and shear wall, with some rigid members connecting the two to represent the rigid floor diaphragm.

The model runs fine, and the stress plots on the quad elements look good, but the model fails statics checks. The global reactions are larger than the sum of the applied forces. The 'check_statics' function built in to the 'analyze' method confirms this.

Any thoughts on what could be causing this?

The code to generate the model is complex so i've pickled the model and attached it in a zip file below:

[stability_test.zip](https://github.com/JWock82/PyNite/files/7234752/stability_test.zip)

I have subclassed PyNite.FEModel3D so you'll need to use the below code to read the model.

```
import pickle

infile = ...

class customUnpickler(pickle.Unpickler):
    def find_class(self, module, name):
        if module == "Analysis.AnalysisBase":
            module = "PyNite.FEModel3D"
            name = "FEModel3D"

        return super().find_class(module, name)
    
with open(infile,mode="rb") as f: model=customUnpickler(f).load()


```

Also, i've noticed that `Visualization._PrepContour` fails when there are nodes in the model not connected to Quad or Plate elements.
`node.contour = sum(node.contour)/len(node.contour)` results in a divide by zero error.

**Comment by JWock82** [comment]
Can you confirm you're using the latest version (0.0.49)? I recently fixed an issue in the statics check related to quadrilateral elements. I want to make sure this is not the same issue.

**Comment by bjhowie** [comment]
Yes, using 0.0.49

**Comment by JWock82** [comment]
OK. I'll have a closer look at this.

**Comment by JWock82** [comment]
Another question. I haven't looked at your model yet. But I did take another look at the code calculating the reactions. I had not updated it to account for spring supports yet. That's a newer feature I recently added. Does your model have spring supports?

**Comment by bjhowie** [comment]
No, no spring supports.

The lowest level of the quad elements have their translations fully fixed, otherwise no other supports.
________________________________

**Comment by JWock82** [comment]
So I've fixed the issue with the divide by zero error. I've also published a new release as v0.0.50 with a few changes to the calculate reactions method. I'm not sure if this fixed it, but I can't check it since the model is pickled. I changed the Node class in version 0.0.50, so I get errors when I run your pickled version. Can you either verify that version 0.0.50 fixes your problem, or send me a new file that's compatible with v0.0.50?

**Comment by bjhowie** [comment]
Contours seem to be working now. Still having issues with the statics though.

I've re-run and attached the pickled model below. You'll still need to use the custom unpickler above to deal with the subclassed model class. 
Let me know if it still doesnt work.

[StabilityModel.zip](https://github.com/JWock82/PyNite/files/7271136/StabilityModel.zip)

**Comment by JWock82** [comment]
I'm trying to narrow the list of possible problems down. Will you confirm a few things for me?

1. Does your model have any quadrilateral surface pressures applied? I didn't see any when I rendered it. It looked like nodal loads only.
2. Are your elements rectangular? If so, can you switch them to rectangular plate elements and see if you have the same issue. That would help me determine if it's a problem with the quad element formulation or something else.

**Comment by bjhowie** [comment]
1. No, node loads only
2. In the model i sent you, they are not perfectly rectangular. I created another model with rectangular plate elements and the issue still seemed to exist, so its likely not related to the quad element formulation

**Comment by bjhowie** [comment]
That makes some sense. I dont think it's the non-orthogonal quads that are doing it though, the problem seems to be with the connection of the beam elements to the quads. If i remove all of the beam elements the issue is resolved, even with the quads at ~45 degrees. Even if there are no loads applied to or via the beam elements it throws off the base reactions

The takeaway then is that, with the current element formulations, beam elements cannot connect to quad elements at a single point? 

Dissapointing outcome. Is there a way that PyNite can at least detect these drilling conditions and throw an exception instead of happily reporting incorrect results?

**Comment by JWock82** [comment]
Also you may try using additional beam elements around the perimeter of the floor attached to the beams and plates. That would resolve the drilling by converting it into a force couple. That’s how my FEA text book suggests handling drilling.

**Comment by bjhowie** [comment]
Thanks. I found the same eng-tips article and thought it was quite helpful.

I'm not sure that this is the full story though. We shouldnt need to consider any drilling degrees of freedom to make this model solve (unlike the flag pole example in the article). There are many other loads paths that can transmit the load via translational push/pull action, but seemingly any model that is capable of attracting any miniscule amount of drilling load to the quads is thrown off entirely. Maybe i'm missing something.

Reducing torsional stiffness of rigid links makes no difference.

I will keep looking into it.

**Comment by JWock82** [comment]
@bjhowie try running this model again with the latest version. I’ve corrected an issue I found with quadrilaterals related to their in-plane stiffness.

**Comment by bjhowie** [comment]
Hey, sorry i never got around to replying to your previous post. Unfortunate outcome, but sounds like it's an inherent limitation of the MITC4 plate formulation.

I ran it again with 0.0.52 but got similar results, in fact the phantom RY reaction increased slightly

![image](https://user-images.githubusercontent.com/45848617/140628575-952d9744-1a59-441c-8390-bdab2a28b933.png)

The Y and Z reactions are small, but maybe not neglegible depending on the application.

**Comment by JWock82** [comment]
Thanks bjhowie. I just added more unit testing for plates/quads to compare them against published solutions. Overall I'm finding that the plates are giving very accurate results for out-of-plane, in-plane, and combined forces.

**Comment by JWock82** [comment]
@bjhowie,
Today I fixed an issue related to statics checks that may resolve this issue. I don't have access to your original file that showcased the issue. Do you still have it handy by chance?

**Comment by bjhowie** [comment]
I should have it somewhere. I'll have a dig around tonight and see if I can find it.

**Comment by bjhowie** [comment]
I couldn't find the code that generated the original model, but I used the pickled data attached in the post above to regenerate the model using version 0.0.93.

Your updates seem to have fixed the issue! The model only has loads applied along the X axis so the below reaction sums look correct to me.
![image](https://github.com/JWock82/PyNite/assets/45848617/a54e3c9e-6f44-490d-82b3-d3cde96644a3)

**Comment by bjhowie** [comment]
I couldn't find the code that generated the original model, but I used the pickled data attached in the post above to regenerate the model using version 0.0.93.

Your updates definitely seem to have helped the issue. RY and RZ are basically zero. I'm unsure if the RMX reaction is an issue? How is it calculated?
![image](https://github.com/JWock82/PyNite/assets/45848617/12f439d7-364a-406a-bded-6802db1c81be)

**Comment by bjhowie** [comment]
OK so i manage to find my old code that generated that model. Doesn't look like your latest changes have helped.

I've used it to generate two more models, one using quad elements and one using plates.
[StabilityModels.zip](https://github.com/JWock82/PyNite/files/14625989/StabilityModels.zip)

This is the statics check for the quads model:
![image](https://github.com/JWock82/PyNite/assets/45848617/a1ffc9a0-79fe-487c-9631-9b0b7f769850)

And for the plates model:
![image](https://github.com/JWock82/PyNite/assets/45848617/ce17a5ad-a4e3-4e69-b1d3-75c12a985c83)

Fairly similar but the phantom reactions change slightly, which is probably not unexpected.

I've got some members in there that are intended to represent a rigid diaphragm. In the examples above ive given them the material properties of steel, but if I increase the E value to something like 1E15 (effectively infinitely stiff) then the phantom reactions get very large.

![image](https://github.com/JWock82/PyNite/assets/45848617/c5e849db-8a8a-4f91-92d3-df7e0eba5786)
